### PR TITLE
Add rename file and move core functionality into core in Angular

### PR DESCRIPTION
**Pull request by kaidohallik**
This PR is related to closed PR #10340. Two PR-s, #10360 and #10492, already implemented some parts of initial #10340.  
I initially canceled finishing #10340 and added https://github.com/jhipster/generator-jhipster/issues/10220#issuecomment-528311220 because one requirement from @vishal423 was not met: Git commit window didn't show that files are renamed.  
After doing #10492 I looked at that again and found out that to fulfill this requirement I need to add basically one additional line of code to #10340: `git mv ${oldFile} ${newFile}`  
So I create this PR with remaining #10340 changes with addition of this `git mv` command. Description of this PR:
* adds functionality to rename files in upgrade process
* this added rename functionality is used to move Angular core classes under core folder

Two images to illustrate output after using this version.

Output from generator console:

regenerate-console

Output from Git commit window:

git-commit

-   Please make sure the below checklist is followed for Pull Requests.

-   [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [ ] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by vishal423** [comment]
I was thinking about leveraging shelljs [move](https://github.com/shelljs/shelljs#mvoptions--source--source--dest) command for this purpose.

**Comment by kaidohallik** [comment]
ShellJS `move` doesn't help here because this doesn't give information to Git that file is moved and Git commit can not bind new and old file location.

**Comment by kaidohallik** [comment]
I would like to make PR to close `Hacktoberfest` labeled issue #10220, but merging this PR is precondition to that.  
@wmarques you approved #10340, this new PR basically adds only one additional line: `git mv` to tell Git that file is moved. What do you think to you have time to review this additional change and merge this PR?  
Or @vishal423 are you interested and have time to review and merge this PR?

**Comment by wmarques** [comment]
@kaidohallik LGTM, however I'm not sure if we can merge this right now as this contains breaking changes. Should we wait for a major release ?

Also, I'm not convinced about the `git mv` usage there, we never did this before.
@jhipster/developers WDYT about this ?

**Comment by kaidohallik** [comment]
@wmarques Why do you think that these are breaking changes? In my opinion these are changes in app internal file locations. I don't believe that these renamed files are imported in some other modules or components than `core.module` - this conclusion tells me that this is not a breaking change.

Do you see any drawbacks using `git mv` command?

File: generators/client/templates/angular/src/main/webapp/app/core/core.module.ts.ejs [hunk 1]
```diff
@@ -20,14 +20,14 @@ import {
...
 
 <%_ if (authenticationType === 'jwt') { _%>
-import { AuthInterceptor } from 'app/blocks/interceptor/auth.interceptor';
+import { AuthInterceptor } from './interceptor/auth.interceptor';
 <%_ } _%>
-import { AuthExpiredInterceptor } from 'app/blocks/interceptor/auth-expired.interceptor';
-import { ErrorHandlerInterceptor } from 'app/blocks/interceptor/errorhandler.interceptor';
-import { NotificationInterceptor } from 'app/blocks/interceptor/notification.interceptor';
+import { AuthExpiredInterceptor } from './interceptor/auth-expired.interceptor';
+import { ErrorHandlerInterceptor } from './interceptor/error-handler.interceptor';
```
Comment (MathieuAA) [inline generators/client/templates/angular/src/main/webapp/app/core/core.module.ts.ejs]: nice

File: generators/generator-base-private.js [hunk 2]
```diff
@@ -1413,4 +1414,77 @@ module.exports = class extends Generator {
...
+        if (extension === '.ts') {
+            options = { parser: 'typescript' };
+        }
+        try {
+            // normalize line endings, because for example in Windows this can be in generated content \n and in disk \r\n
+            return prettierFormat(sourceCode, options)
+                .replace(new RegExp('\\r\\n', 'g'), '\n')
+                .replace(new RegExp('\\r', 'g'), '\n');
+        } catch (e) {
+            return sourceCode.replace(new RegExp('\\s', 'g'), '');
```
Comment (MathieuAA) [inline generators/generator-base-private.js]: why don't you use this form? `replace(/WHATEVER/g, '\n')`

**Comment by MathieuAA** [comment]
Why do you want to use `git mv`? I'm not for or against it, I just don't understand the need to.

**Comment by vishal423** [comment]
While working on a new prototype, I have noticed that starting `v7`, angular style guide doesn't mention about `core` module. Purpose of `core` module recommendation was to ensure singleton services, that got achieved with `providedIn`.

File: generators/generator-base-private.js [hunk 3]
```diff
@@ -1413,4 +1414,77 @@ module.exports = class extends Generator {
...
+    /**
+     * Formats source code so for compare
+     * @param {string} sourceCode - source code to format
+     * @param {string} extension - extension for configuring Prettier properly, for example `.ts` for Typescript files
+     */
+    formatSourceCodeForCompare(sourceCode, extension) {
+        let options = {};
+        if (extension === '.ts') {
+            options = { parser: 'typescript' };
+        }
```
[the remark this answers is not in the document]
Comment (kaidohallik) [reply generators/generator-base-private.js]: Thanks, I made this change.

File: generators/client/index.js [hunk 4]
```diff
@@ -148,6 +149,8 @@ module.exports = class extends BaseBlueprintGenerator {
                         this.clientPackageManager = 'yarn';
                     }
                 }
+                // preserve old jhipsterVersion value for cleanup which occurs after new config is written into disk
```
[the remark this answers is not in the document]
Comment (kaidohallik) [reply generators/client/index.js]: OK, removed disk part.

File: generators/generator-base-private.js (hunk 2 above)
Comment (kaidohallik) [reply generators/generator-base-private.js]: The `new RegExp` command was just first one that came to my mind. Your suggestion is of course cleaner, shorter and better readable. I made this change. Thanks!

**Comment by kaidohallik** [comment]
> Why do you want to use `git mv`? I'm not for or against it, I just don't understand the need to.

The goal to be achieved is: if user upgrades existing app then user must see in Git commit preview that files are renamed.  
`git mv` is exactly the command that does it, I don't see any other solution to fulfill this aim.

**Comment by kaidohallik** [comment]
> While working on a new prototype, I have noticed that starting `v7`, angular style guide doesn't mention about `core` module. Purpose of `core` module recommendation was to ensure singleton services, that got achieved with `providedIn`.

Currently core module performs services and third party modules configuration. I think that splitting services and third party modules configuration into separate module makes code better structured.  
What can be alternatives:
* move all content from `core.module` to `app.module` and remove `core.module`?
* rename `core.module` to `core-libs.module`?

I prefer to keep `core.module` as this is ATM and if someone has better idea then maybe discuss that in a new issue?

**Comment by vishal423** [comment]
I would like to discuss module restructuring in separate issue that can be targeted for `v7`. IMO, we should try to align our generated code with angular style guide. Unless there is a real issue being targeted here, I don't see an immediate need to move these files from `app` to `core` module.

**Comment by kaidohallik** [comment]
OK, let's close this for now and discuss in the new issue.

**Pull request closed without merging**